source/common/http/conn_manager_impl.cc

```diff
@@ -1120,10 +1121,19 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(RequestHeaderMapSharedPt
     request_header_timer_.reset();
   }
 
-  // Both saw_connection_close_ and is_head_request_ affect local replies: set
-  // them as early as possible.
+  // Both shouldDrainConnectionUponCompletion() and is_head_request_ affect local replies: set them
+  // as early as possible.
   const Protocol protocol = connection_manager_.codec_->protocol();
-  state_.saw_connection_close_ = HeaderUtility::shouldCloseConnection(protocol, *request_headers_);
+  if (Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.http1_connection_close_header_in_redirect")) {
+    if (HeaderUtility::shouldCloseConnection(protocol, *request_headers_)) {
+      filter_manager_.streamInfo().setShouldDrainConnectionUponCompletion(true);
+    }
```

Comment:
I'm not sure I'm right. Is it possible to determine this request is `redirect` from the stream info? (or we add some flag to the stream info). Then the code will self-explain better.

Otherwise it would be great to add some comments at here to explain why we didn't override the `shouldDrainConnectionUponCompletion` when it is false here.